qiskit_optimization/applications/sk_model.py

```diff
@@ -0,0 +1,108 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""An application class for the Sherrington Kirkpatrick (SK) model."""
+
+
+from typing import Union, List
+
+import networkx as nx
+import numpy as np
+from docplex.mp.model import Model
+
+from .graph_optimization_application import OptimizationApplication
+from ..algorithms import OptimizationResult
+from ..problems.quadratic_program import QuadraticProgram
+from ..translators import from_docplex_mp
+
+
+class SKModel(OptimizationApplication):
+    r"""Optimization application of the "Sherrington Kirkpatrick (SK) model" [1].
+
+    The SK Hamiltonian over n spins is given as:
+        :math:`H(x)=-1/\sqrt{n} \sum_{i<j} w_{i,j}x_ix_j`, :math:`where x_i\in\{\pm 1\}`
+    and :math:`w_{i,j}\in\{\pm 1\}` are chosen independently and uniformly at random.
+    Notice that there are other variants e.g., with :math:`w_{i,j}` chosen from the normal distribution
+    with mean 0 and variance 1.
+
+    References:
+        [1]: Dmitry Panchenko. "The Sherrington-Kirkpatrick model: an overview",
+        https://arxiv.org/abs/1211.1094
+    """
+
+    def __init__(self, num_of_sites: int, rng: np.random.RandomState = np.random.RandomState(0)):
+        """
+        Constructor for the SK model.
+
+        Args:
+            num_of_sites: number of sites
+            rng: numpy pseudo-random number generator
+        """
+        self._rng = rng
+        self._num_of_sites = num_of_sites
+        self._graph = nx.complete_graph(self._num_of_sites)
+
+        self.new_instance()
+
+    def new_instance(self) -> None:
```

Comment:
I have a concern about this public method. Users can generate different SK model by giving different values to the constructor. Why do you think this method is necessary no top of the constructor? @X-Libor

Reply:
In general you are interested in typical behavior for this model. Hence, you are going to generate large number of instances of the same number of sites with different --- random --- distortions.

Maybe the whole class is more a factory for SK model instances.

Reply:
You don't need to overwrite an existing object to try another instance. I suggest to generate a new object with a new random number generator as suggested https://github.com/Qiskit/qiskit-optimization/pull/208#discussion_r677635115

Reply:
then you would need to keep the rng somewhere else to have it reproducible.
 :(

I also dont want that some other part of code can change the state of the rng and make it not reproducible.

Reply:
I guess I am still not clear. When you say 100000 instances these are all different I assume - but reproducible as a set given the same random number generator. Hence you want the same generator to be used by different instances otherwise if they have their own generator and its seeded they will all be the same assuming the same seed. Of course you can do with 100000 different fixed seed values - but maybe thats too much. On the other hand how many problem instances do you normally expect users to play with?

Reply:
Yes, in that example I meant 100000 random instances. Reproducible: they would be generated by one independent seeded random number generator. Now they are produced by invoking the new_instance method, which changes the weights.

Maybe this should be implemented through some design pattern like factory.

Reply:
If you are allowed to destroy an existing object by `new_instance` as `sk_model.new_instance()`, it is essentially same as replacing an object with a new object with a different edge weights as `sk_model = SKModel(new_edge_weights)`. Generating an object from edge weights is intuitive from reproducibility point of view.

Reply:
I dont want users to pick edge weights. They would possibly not do it in the way model is intended.

Reply:
What if users want to solve a SK model generated by their own rng? I believe we need to make edge weights as an input of the constructor and provide a utility static method to generate an instance with numpy's rng similar to `new_instance`.

Reply:
If maxcut is sufficient to represent the problem, we don't need to provide a new class. It seems enough to make a `generate_sk_model` to `Maxcut` class.

Reply:
> What if users want to solve a SK model generated by their own rng?
Ok. Nevertheless if they want this to be repeatable then it has to be an independent rng.

You mention in the docstring that
> 
> > Notice that there are other variants e.g., with :math:`w_{i,j}` chosen from the normal distribution
This is not related to the previous.

> So, some users may want to set edge weights generated from a normal distribution.
Yes, it can be generalized in some other iteration. Right now this is about models with weights \pm 1 with uniform independent.
 
> I believe we need to make edge weights as an input of the constructor and provide a utility static method to generate an instance with numpy's rng similar to `new_instance`.
:( no, then the model would not necessarily have the properties. It is solvable model if people will put something else in it it would not be that model. (for more info about the SK model please see: https://arxiv.org/abs/1211.1094)
Ultimately we wont this to have these properties.

Reply:
As for `new_instance`, I think it's better to make it a static method as we do in other classes. To try many instances the following codes are essentially same.
```python
# current impl
meo = MinimumEigenOptimizer(...)
mod = SKModel(4, np.random.default_rng(1))
for i in range(n):
    qp = mod.to_quadratic_program()
    result = meo.solve(qp)
    ... process results
    mod.new_instance()
``` 

```python
# my proposal
meo = MinimumEigenOptimizer(...)
for i in range(n):
    mod = SKModel.random_instance(4, np.random.default_rng(i))
    qp = mod.to_quadratic_program()
    result = meo.solve(qp)
    ... process results
``` 

To reproduce k-th SK model, the first code needs to iterate k times. But the second one can generate k-th model directly.
```python
# current impl
mod = SKModel(4, np.random.default_rng(1))
for i in range(k-1):
    mod.new_instance()

# my proposal
mod = SKModel.random_instance(4, np.random.default_rng(k))
```

Reply:
I am not clear why `mod = SKModel(4, np.random.default_rng(k))` in "current implementation"
does not do what you propose?

Reply:
Users may want to manually set edge weight either 1 or -1. For example, they may want to flip an edge weight and solve it. If you need to ensure all edge weights are ±1, you can check it in the constructor and raise a warning if not satisfied.

Reply:
You make an instance `problem = SKModel(self._num_of_sites, np.random.RandomState(self._seed))` in the unit test. The difference is not essential. I put a random number generator while you use legacy random state.
https://numpy.org/doc/stable/reference/random/generator.html
https://numpy.org/doc/stable/reference/random/legacy.html

Reply:
My question is suppose you try  n different `SKModel`s and how do you get the n-1 th model? Is it possible to directly obtain one using `new_instance`?

Reply:
I maybe got your idea. If you need to reproduce a model you use the constructor; and if you don't have to reproduce any model, you use `new_instance` for efficiency. Is it correct?

Reply:
> Users may want to manually set edge weight either 1 or -1. For example, they may want to flip an edge weight and solve it. If you need to ensure all edge weights are ±1, you can check it in the constructor and raise a warning if not satisfied.

If they put there their own weights they could do it in a way that does not conform with the model. something which has no mean 0 and variance 1.

Reply:
> I maybe got your idea. If you need to reproduce a model you generate models with the constructor; and if you don't have to reproduce any model, you use `new_instance` to generate models for efficiency. Is it correct?

I dont understand. if you generate new instances it is reproducible if your pseudo random number generator is independent.

Use case: what is the expected energy  2-layer qaoa achieves for SK model with 20 sites. Does the variational parameters concentrate around some value, and so.

Reply:
> I follow your unit test `problem = SKModel(self._num_of_sites, np.random.RandomState(self._seed))`. The difference is not essential. I put a random number generator while you use legacy random state. Why do you want to use a legacy one?
> https://numpy.org/doc/stable/reference/random/generator.html
> https://numpy.org/doc/stable/reference/random/legacy.html

https://github.com/Qiskit/qiskit-optimization/pull/208#discussion_r675874389